db/collection.go

```diff
@@ -229,24 +207,38 @@ func (db *db) updateCollection(
 	ctx context.Context,
 	txn datastore.Txn,
 	existingDescriptionsByName map[string]client.CollectionDescription,
-	proposedDescriptionsByName map[string]client.CollectionDescription,
+	existingSchemaByName map[string]client.SchemaDescription,
+	proposedDescriptionsByName map[string]client.SchemaDescription,
```

Comment:
thought(out of scope): I would like to point out that this in a really confusing set of params. We are updateing one collection but we are passing all collection, all schemas and all proposed schemas. In my opinion, only what is directly relevant should be passed to this function.

Reply:
This one is fiddly, and I agree that it is a bit of an eyesore. However they are all used and all need to be there (or within this func, we could probably (re)fetch them) due to relational fields.

It does get improved slightly in a later PR when we are able to change all their types to SchemaDescriptions (allowing us to only pass in two maps instead of 3).  Might be worth re-visiting this in https://github.com/sourcenetwork/defradb/pull/1965

Reply:
I've added a todo in that PR, as one of the params still makes no sense there: https://github.com/sourcenetwork/defradb/pull/1965/files#r1358767728

Removing it leaves us with:
```
func (db *db) updateCollection(
	ctx context.Context,
	txn datastore.Txn,
	existingSchemaByName map[string]client.SchemaDescription,
	proposedDescriptionsByName map[string]client.SchemaDescription,
	schema client.SchemaDescription,
	setAsDefaultVersion bool,
) (client.Collection, error) {
```
`existingSchemaByName` could be re-fetched inside `updateCollection`, but that feels fairly wasteful atm (although a planned repository mechanic would remove that issue).

Reply:
They are used but in a wasteful way as we only care about one in each of the maps. The only proposedDescription that we care about is already passed as `schema` so that one can be removed completely. It should probably look like this:
```go
func (db *db) updateCollection(
    ctx context.Context,
    txn datastore.Txn,
    existingSchema client.SchemaDescription,
    schema client.SchemaDescription,
    desc client.CollectionDescription
    setAsDefaultVersion bool,
) (client.Collection, error) {
```

Although the checks on the schema should probably be done before calling `updateCollection` and we would end up with just:
```go
func (db *db) updateCollection(
    ctx context.Context,
    txn datastore.Txn,
    schema client.SchemaDescription,
    desc client.CollectionDescription
    setAsDefaultVersion bool, // if we call updateCollection, this should be implicit and removed from params.
) (client.Collection, error) {
```

Reply:
> The only proposedDescription that we care about is already passed as schema so that one can be removed completely

This is not true, we must validate that all relational fields point to a valid schema.  This schema is rarely the one being updated.

> Although the checks on the schema should probably be done before calling updateCollection

I'me very much against that, as it moves the validation out of a request-language agnostic part of the codebase (db.collection.go) into a request-language specific area (patchSchema) of the codebase.  It would mean that if we add an alterative to patchSchema (including an embedded Go call) we'd have to rework the validation.

Reply:
Maybe the problem then is that the name of the function is `updateSchema` put it handles both schema and collection in the same function. Maybe once we get further in cleaning this up it won't be a problem anymore.

Reply:
for-reference: We spoke about this last Friday, and there is a task on a later PR to rename this function (when it stops being updateCollection)